core_impl/src/main/java/io/opencensus/trace/internal/ConcurrentIntrusiveList.java

```diff
@@ -0,0 +1,185 @@
+/*
+ * Copyright 2017, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opencensus.trace.internal;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import io.opencensus.trace.internal.ConcurrentIntrusiveList.Element;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * An {@code ConcurrentIntrusiveList<T>} is a doubly-linked list where the link pointers are
+ * embedded in the elements. This makes insertion and removal into a known position constant time.
```

Comment:
I'd prefer things like this introduced when they are used as it helps better tell the story of what it is doing. For example, benchmarks could show if it works well for the use case.

One thing interesting about this is mention of a "known position", though the operations on the list aren't positional in nature (or at least not explicitly so). indeed Element is positional, so this hints at how the list is used.. which is to my point about it being clearer if this is raised when it is used :P